std/algorithm/comparison.d

```diff
@@ -1567,3 +1567,146 @@ unittest
                 2, "two",
                 ));
 }
+
+/** Returns: First Argument (element of $(D a)) whose implicit conversion to
+    $(D bool) is $(D true.
+
+    Similar to behaviour of or operator in dynamic languages such as Lisp's (or
+    a...) and Python's a or ....
+
+    TODO: Is inout Conversion!T the correct return value?
+
+    NOTE: Lazy parameters are currently inferred by DMD to alwasy throw. This
+    makes it impossible to mark $(D either) as $(D nothrow). Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(Ts...)(lazy Ts a)
+if (a.length >= 1 && !is(CommonType!Ts == void))
+{
+    foreach (e; a)
+        if (e)
+            return e;
+    return CommonType!Ts.init;
+}
+
+/// ditto
+auto ref either(Ts...)(ref Ts a)
+if (a.length >= 1 && allSameType!Ts)
+{
+    foreach (ref e; a[0 .. $ - 1])
+        if (e)
+            return e;
+    return a[$ - 1];
```

Comment:
IMO the important thing here is to have the same behaviour for l-values as for r-values. If so I see three alternatives here:

1. Mimic Python's behaviour;
- `either` returns *last* argument upon failure
- `every` returns *first* argument upon failure

2. If default value is supplied return it upon failure for both l-value and r-value overloads.

4. Throw Exceptions

What do you think?